src/cdk-experimental/testing/component-harness.ts

```diff
@@ -0,0 +1,190 @@
+/**
+ * @license
+ * Copyright Google LLC All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import {TestElement} from './test-element';
+
+/** Options that can be specified when querying for an Element. */
+export interface QueryOptions {
+  /**
+   * Whether the found element can be null. If allowNull is set, the searching function will always
+   * try to fetch the element at once. When the element cannot be found, the searching function
+   * should return null if allowNull is set to true, throw an error if allowNull is set to false.
+   * If allowNull is not set, the framework will choose the behaviors that make more sense for each
+   * test type (e.g. for unit test, the framework will make sure the element is not null; otherwise
+   * throw an error); however, the internal behavior is not guaranteed and user should not rely on
+   * it. Note that in most cases, you don't need to care about whether an element is present when
+   * loading the element and don't need to set this parameter unless you do want to check whether
+   * the element is present when calling the searching function. e.g. you want to make sure some
+   * element is not there when loading the element in order to check whether a "ngif" works well.
+   */
+  allowNull?: boolean;
+  /**
+   * If global is set to true, the selector will match any element on the page and is not limited to
+   * the root of the harness. If global is unset or set to false, the selector will only find
+   * elements under the current root.
+   */
+  global?: boolean;
+}
+
+/** Interface that is used to find elements in the DOM and create harnesses for them. */
+export interface HarnessLocator {
+  /**
+   * Get the host element of locator.
+   */
+  host(): TestElement;
+
+  /**
+   * Search the first matched test element.
+   * @param selector The CSS selector of the test elements.
+   * @param options Optional, extra searching options
+   */
+  querySelector(selector: string, options?: QueryOptions): Promise<TestElement|null>;
+
+  /**
+   * Search all matched test elements under current root by CSS selector.
+   * @param selector The CSS selector of the test elements.
+   */
+  querySelectorAll(selector: string): Promise<TestElement[]>;
```

Comment:
Should `querySelectorAll` accept `options` as well?

Reply:
I'm planning to address this when I make some API changes in the next PR. I think this is currently implemented like this because `allowNull` doesn't really make sense for `querySelectorAll`